security/pkg/nodeagent/caClient/client.go

```diff
@@ -37,27 +43,45 @@ type caClient struct {
 }
 
 // NewCAClient create an CA client.
-func NewCAClient(addr string, dialOptions []grpc.DialOption) (Client, error) {
-	conn, err := grpc.Dial(addr, dialOptions...)
+func NewCAClient(endpoint, rootPath string) (Client, error) {
+	var opts grpc.DialOption
+	if rootPath != "" {
+		creds, err := credentials.NewClientTLSFromFile(rootPath, "")
+		if err != nil {
+			log.Errorf("Failed to read root certificate file: %v", err)
+			return nil, fmt.Errorf("failed to read root cert")
+		}
+		opts = grpc.WithTransportCredentials(creds)
+	} else {
+		opts = grpc.WithInsecure()
+	}
+
+	conn, err := grpc.Dial(endpoint, opts)
 	if err != nil {
-		log.Errorf("Failed to connect to CA: %v", err)
-		return nil, err
+		log.Errorf("Failed to connect to endpoint %q: %v", endpoint, err)
+		return nil, fmt.Errorf("failed to connect to endpoint %q", endpoint)
 	}
 
 	return &caClient{
 		client: capb.NewIstioCertificateServiceClient(conn),
 	}, nil
 }
 
-func (cl *caClient) CSRSign(ctx context.Context, csrPEM []byte, subjectID string,
+func (cl *caClient) CSRSign(ctx context.Context, csrPEM []byte, token string,
 	certValidTTLInSec int64) ([]byte /*PEM-encoded certificate chain*/, error) {
+	// TODO(quanlin): remove SubjectId field once CA put it as optional.
+	subjectID, err := parseSubjectID(token)
+	if err != nil {
+		return nil, errors.New("failed to construct IstioCertificate request")
```

Comment:
Put err into this error info as well? Otherwise we will lose the detailed parseSubjectID error info.

Reply:
ok